File: src/TspJulia/elements/Elements.jl
module Elements

include("point.jl")
include("route.jl")
include("cities.jl")

export Point, Cities, Route, distance, compute_distance_matrix, total_distance

"""
Distance function.
"""
distance(p1::Point, p2::Point) = √((p1.x - p2.x)^2 + (p1.y - p2.y)^2)

"""
Compute distance matrix.
"""
function compute_distance_matrix(cities::Cities)
    return [
        distance(cities.cities[i], cities.cities[j]) for i ∈ 1:length(cities.cities),
        j ∈ 1:length(cities.cities)
    ]
end

function total_distance(cities::Cities, route::Route)
    #! format: off
    return (
        sum(
            distance(cities[route[i]], cities[route[i + 1]])
            for i ∈ 1:(length(route) - 1)
        )
        + distance(cities[route[end]], cities[route[1]])
    )
    #! format: on
end

end  # module Elements

File: src/tsp/py/elements.py

import typing as t
from collections import deque
from dataclasses import dataclass, field
from functools import cached_property

import numpy as np

from ..utils import distance_matrix

__all__ = ["Point", "Cities", "Route"]

@dataclass
class Point:
    x: float
    y: float

@dataclass
class Route:
    route: list[int]

    def __len__(self) -> int:
        return len(self.route)

    def __getitem__(self, item) -> int:
        return self.route[item]
    
    def __iter__(self):
        return iter(self.route)
        
    def copy(self) -> "Route":
        return Route(list(self.route))

    def to_jl(self):
        # Helper for compatibility if needed, though this is pure python impl
        return self.route

@dataclass
class Cities:
    """
    Wrapper for points. Distance matrix is computed lazily.
    """
    points: list[Point]
    _dist_matrix: t.Optional[np.ndarray] = field(default=None, repr=False, init=False)

    @cached_property
    def dist_matrix(self) -> np.ndarray:
        # Convert points to array Nx2
        coords = np.array([[p.x, p.y] for p in self.points])
        return distance_matrix(coords)

    def __len__(self) -> int:
        return len(self.points)

    def total_distance(self, route: Route) -> float:
        """Compute the total distance of the route."""
        # Logic from original sa.py
        r = route.route
        rot_route = deque(r)
        rot_route.rotate(-1)
        
        matrix = self.dist_matrix
        return sum([matrix[i, j] for i, j in zip(r, rot_route)])
